alf/trainers/policy_trainer.py

```diff
@@ -672,6 +677,77 @@ def _train(self):
+        """Periodically check the consistency of model parameters of different
+        DDP processes.
+
+        Note that DDP can only make sure that the parameters requiring gradients
+        are always consistent across processes, but cannot guarantee the
+        same thing for those without gradients. An example scenario is the target
+        model of SAC.
+
+        Here we check the parameters for both cases. If any inconsistency is found,
+        a warning message will be printed, without interrupting the training.
+
+        Even if all model parameters require gradients, this function can still
+        serve as a sanity checker for our DDP implementation.
+
+        Args:
+            iter_num: current training iteration
+            training_started: if training has started or not. ALF only wraps DDP
+                around algorithms after first training forward, when model parameters
+                are first synced. So only consistency checking after this is meaingful.
+                Note that for off-policy training before first gradient update,
+                different processes might use different models to unroll.
+        """
+        if not training_started:
+            return
+
+        proc_cxt = PerProcessContext()
+        if (proc_cxt.is_distributed
+                and self._config.ddp_paras_check_interval > 0
+                # Assume that DDP will make sure that this modulo check won't
+                # cause a dead lock, i.e., all workers have the same ``iter_num``
+                # at any moment.
+                and iter_num % self._config.ddp_paras_check_interval == 0):
+
+            with alf.summary.record_if(lambda: True):
+                with record_time("time/para_consistency"):
+
+                    paras_stat = self._algorithm.compute_paras_statistics()
+                    queue = proc_cxt.paras_queue
+                    if self._rank > 0:
+                        # Put para stat into the queue. Don't need to wait rank=0's
+                        # return, because DDP will sync processes at the next
+                        # gradient update.
+                        queue.put(paras_stat)
+                    else:
+                        consistent = True
+                        # rank=0 gets all other para stats
+                        for i in range(proc_cxt.num_processes - 1):
+                            their_paras_stat = queue.get()
+                            is_close = np.isclose(
+                                paras_stat, their_paras_stat, atol=1e-6)
+                            if not np.all(is_close):
+                                consistent = False
+                                common.warning(
+                                    "Found inconsistent parameter stats across "
+                                    "DDP processes: %s vs. %s => %s" %
+                                    (paras_stat, their_paras_stat, is_close))
```

Comment:
should include the name of the parameter. otherwise it will be very hard to know the problem

Reply:
> should include the name of the parameter. otherwise it will be very hard to know the problem

Currently the stats is computed for the entire parameter space. Are you suggesting that we compute the stats for each named parameter? 

(Because the inconsistency will diffuse very quickly during gradient update, I wonder if providing parameter names helps).

Reply:
The communication cost of a dict of param name to (mean, std) shouldn't be that much, especially if the interval is large enough.

I think in most situations, the inconsistency will not spread. If a parameter is updated by an optimizer, all the machines will use the same gradient to update it, so it should remain consistent, even if some other parameters become inconsistent. The inconsistency only comes from some special logic, which can be easier to be identified if the name of the parameter is known.

Reply:
> The communication cost of a dict of param name to (mean, std) shouldn't be that much, especially if the interval is large enough.
> 
> I think in most situations, the inconsistency will not spread. If a parameter is updated by an optimizer, all the machines will use the same gradient to update it, so it should remain consistent, even if some other parameters become inconsistent. The inconsistency only comes from some special logic, which can be easier to be identified if the name of the parameter is known.

Right, all parameters being updated will always be the same.